Discussion optimization (#46)
opendihu/optimization: BayesianOptimization/discussion_parameters.md

> ## Options for Bayesian Optimization

Reviewer: These are specific of botorch right? If yes, mention it and add a link to the botorch documentation.

> Bayesian Optimzation can optimize very diverse functions depending on its parameters. These parameters can be adjusted in the file ```setup_BayesOpt_general_1D.py```. 

Reviewer: ```suggestion
Bayesian Optimization can optimize very diverse functions depending on its parameters. These parameters can be adjusted in the file ```setup_BayesOpt_general_1D.py```. 
```

- The options for the kernel are the RBF-kernel and the Matérn kernel with $\nu\in${0.5, 1.5, 2.5}.
- The possible mean functions are the constant mean and the zero mean.

> - The noise can be chosen fixed or variable.

Reviewer: Maybe constant sounds better than fixed? 
```suggestion
- The noise can be set to constant or variable.
```

- The implemented acquisition functions are the Expected Improvement (EI), Probability of Improvement (PI), Knowledge Gradient (KG) and Entropy Search (ES).
- The rest are minor changes like the stopping criterion or the initial values.

## General recommendation

> The discussion about the best parameter setup for every case can be complicated and there is generally no best option. There has been a Bachelor thesis that discusses the best parameters for nice functions. The corresponding GitHub repository can be found [here](https://github.com/opendihu/optimization/releases/tag/Bachelor-thesis). The general recommendation of this Bachelor thesis is the Matérn kernel with $\nu$=0.5, the constant mean function, fixed noise, the ES acquisition function and the xy-stopping criterion. But this recommendation is just for the general case of nice functions. If we have a special function of which we have some information, there might be a better option.

Reviewer: ```suggestion
The optimal parameter setup is different for every case, and it is difficult to provide a general recommendation. 
A tentative recommendation is provided in  Lukas Bauer's bachelor thesis (see [tag](https://github.com/opendihu/optimization/releases/tag/Bachelor-thesis)): Matérn kernel with $\nu$=0.5, the constant mean function, fixed noise, the ES acquisition function and the xy-stopping criterion. 
```

Reviewer: @Lukuas31415 is your thesis online? Maybe we could add a link to the written report (in addition to the code tag).

Author: My Bachelor thesis is not online, no. Should I upload it? Where would I do that?

Reviewer: Typically the university does it for you, but I guess it is different for different faculties. What you could do is ask your colleagues from maths to see if they also don't have it online. In any case, we can just close this point for now and if you later find out that it is accessible online add the link in a different PR.

## Biceps case

> There is the [isometric biceps case](../opendihu_examples/isometric_contraction/biceps_muscle/) for which we want to discuss special parameters. If we use the recommended parameters, we get the following optimization plot: 

Reviewer: ```suggestion
Next we discuss the parameters setup for the [isometric biceps case](../opendihu_examples/isometric_contraction/biceps_muscle/). If we use the parameters according to the general recommendation, we get the following optimization plot: 
```

![](../figures/isometric_biceps.png)

We have found a maximum this way, which is good, but there are some disadvantages of this optimization process.

> - There are very large areas of high uncertainty in which we don't know if we might have a hidden maximum

Reviewer: ```suggestion
- There are very large areas of high uncertainty in which we do not know if we might have a hidden maximum
```

> - The mean function is not very smooth and the trials are in local maxima, which is not very realistic

Reviewer: ```suggestion
- The mean function is not smooth and has local maxima at the trial points, which looks suspicious
```

#### Parameters with low impact
There are a few parameters that do not have a lot of impact on the general result, which we are not going to look at here. They can be adjusted after the major parameter choices for finetuning. Examples for that are the mean function, the noise or the stopping criterion. All these parameters will be chosen as in the general recommendation.

#### Entropy Search
The recommended setup with the Matérn kernel and $\nu$=0.5 is aparently working fine, but not optimally. If we change $\nu$ to 1.5 or 2.5, the result is immediately much worse. Since $\nu$ is the smoothness parameter, a higher value changes the assumed smoothness of the function. This has two disadvantages. The first one is the compatibility with ES. This acquisition function is working with minimizing entropy, which is much easier if we have more uncertainty. We lose that with a higher smoothness parameter. The second disadvantage is that we might not find the maximum in this case. The maximum is just at the edge of a high jump downwards. With a high smoothness the mean function will interpolate the trials in a way that will not allow high jumps, so we might look for a maximum elsewhere.

> However, there is the option of adjusting the ES acquisition function so that we at least remove the large areas of high uncertainty. An option for this would be an evaluation of our function every 3 or 4 evaluations at the value with the highest uncertainty. In the case above this would add about 2 evaluations. That can have two effects. Either this results in a new and better maximum, or we get a nicer picture in the end. Keeping in mind that an evaluation takes an hour, one has to evaluate if this is worth the risk. Since you can always do just one more evaluation after the optimization process is done, the acquisition function should probably not be changed (if we're using ES), and depending on the result we can add more evaluations in areas we deem necessary.

Reviewer: ```suggestion
However, there is the option of adjusting the ES acquisition function so that we at least remove the large areas of high uncertainty. An option for this would be an evaluation of our function every 3 or 4 evaluations at the value with the highest uncertainty. In the case above this would add about 2 evaluations. That can have two effects. Either this results in a new and better maximum, or we get a nicer picture in the end. Keeping in mind that an evaluation takes an hour, one has to evaluate if this is worth the risk. Since you can always do just one more evaluation after the optimization process is done, the acquisition function should probably not be changed (if we are using ES), and depending on the result we can add more evaluations in areas we deem necessary.
```

#### Expected Improvement

> Pairing the EI acquisition function with the RBF-kernel or the Matérn kernel with $\nu$=2.5, we get the same problem as before, that a high jump of the function is not expected and we therefore look for the maximum in the wrong places. For $\nu\in$ {0.5,1.5} we get a different behaviour. With $\nu$=0.5 the uncertainty in a small area around evaluation points is already quite high. That's why the EI tends to expect the next possible maximum closely around already evaluated points instead of looking for it elsewhere. EI tends to choose low risk low reward more than e.g. ES. Therefore the combination with $\nu$=0.5 is not recommended. The combination of EI and $\nu$=1.5 is generally working very well, because the low uncertainty around evaluated points encourages the acquisition function to explore more, but still allows some level of unsmoothness, which is necessary for high jumps. A typical plot of this combination would be the following:

Reviewer: ```suggestion
Pairing the EI acquisition function with the RBF-kernel or the Matérn kernel with $\nu$=2.5, we get the same problem as before, that a high jump of the function is not expected and we therefore look for the maximum in the wrong places. For $\nu\in$ {0.5,1.5} we get a different behaviour. With $\nu$=0.5 the uncertainty in a small area around evaluation points is already quite high. That is why the EI tends to expect the next possible maximum closely around already evaluated points instead of looking for it elsewhere. EI tends to choose low risk low reward more than e.g. ES. Therefore the combination with $\nu$=0.5 is not recommended. The combination of EI and $\nu$=1.5 is generally working very well, because the low uncertainty around evaluated points encourages the acquisition function to explore more, but still allows some level of unsmoothness, which is necessary for high jumps. A typical plot of this combination would be the following:
```

![](../figures/EI_matern_1.5.png)

As we can see, there are no large areas with high uncertainty where there could be a hidden maximum. Also the mean function is very smooth. These are the critique points at the general recommendation, so this combination is better than that in this regard. But there are also disadvantages. For one, there are a lot of evaluations needed, about 3 more than the general recommendation. Also, the found maximum is not very accurate. That happens because of the low uncertainty around the found maximum, which tells the acquisition function, that it's not very likely that there might be a higher value there.

#### Knowledge Gradient

> This acquisition function is extremely computationally expensive. Therefore this is not recommended.

Reviewer: ```suggestion
This acquisition function is not recommended because it is extremely computationally expensive. 
```

Reviewer: Why though?

Author: I can't really test how this works because it takes ages. In comparison to the biceps it probably adds less than an extra evaluation, but I can't guarantee that. And if you want to optimize faster things, this is slower than the evaluation itself. If we get a surrogate model this would be the case. But if you want, I can start looking into KG as well, but that is a slow process.

Author: It doesn't increase the number of evaluations, but calculating the maximum of the acquisition function is extremely expensive. With most of the acquisition functions this is negligible because of the high computational cost of the function evaluation, but not with this one.

#### Probability of Improvement

> With every kernel this acquisition function has the same problem. It doesn't choose any risk at all, but also doesn't get any reward. It always sticks close to the highest already evaluated point and therefore doesn't find the maximum in a reasonable amount of evaluations.

Reviewer: doesn't choose? I don't know what you mean. 
Please write does not instead of doesn't;)

#### Conclusion

> Depending on how important the critique points of large areas of high uncertainty and an unsmooth mean function are to the user, we have three possible rocommendations. Either we stick with the general recommendation and ignore the critique points, or we at least use the modified ES acquisition function to get rid of the large areas of high uncertainty. If the critique points are very important, the recommendation is the EI acquisition function with Matérn kernel and $\nu$=1.5.

Reviewer: ```suggestion
Depending on how important the critique points of large areas of high uncertainty and an non-smooth mean function are to the user, we have three possible recommendations. Either we stick with the general recommendation and ignore the critique points, or we use the modified ES acquisition function to reduce the large areas of high uncertainty. If the critique points are very important, the recommendation is the EI acquisition function with Matérn kernel and $\nu$=1.5.
```
